File: backend/app/modules/sales/service.py
from app.core.constants import SALES_ORIGINS
from app.core.errors import AppError
from app.core.formatters import canonical_text_key, normalize_text
from app.core.validators import ensure_not_future, ensure_positive_number, parse_date, require_fields, validate_enum
from app.modules.attempts.service import AttemptService
from app.modules.leads.service import LeadService
from app.services.base_service import BaseService


class SaleService(BaseService):
    table_name = "vendas"

    def __init__(self):
        super().__init__()
        self.lead_service = LeadService()
        self.attempt_service = AttemptService()

    def list_sales(self, filters: dict) -> list[dict]:
        sql = "select * from vendas where 1=1"
        params: list = []

        if filters.get("origem_fechamento"):
            sql += " and origem_fechamento = %s"
            params.append(filters["origem_fechamento"])
        if filters.get("id_servico"):
            sql += " and id_servico = %s"
            params.append(filters["id_servico"])
        if filters.get("periodo_inicio"):
            sql += " and data_venda >= %s"
            params.append(filters["periodo_inicio"])
        if filters.get("periodo_fim"):
            sql += " and data_venda <= %s"
            params.append(filters["periodo_fim"])

        sql += " order by data_venda desc, id_venda desc"
        return self.db.fetch_all(sql, params)

    def create_sale(self, payload: dict, user_id: str) -> dict:
        require_fields(payload, ["id_lead", "id_servico", "origem_fechamento", "valor_venda", "data_venda"])
        lead = self.lead_service.get_lead(payload["id_lead"])
        self.lead_service.ensure_allows_new_related_records(lead)
        if not self.attempt_service.has_attempt_for_lead(lead["id_lead"]):
            raise AppError("Só é possível registrar venda após existir tentativa de contato.")

        service = self.db.fetch_optional("select * from servicos where id_servico = %s limit 1", [payload["id_servico"]])
        if not service:
            raise AppError("Serviço informado não existe.")

        sale_date = parse_date(payload["data_venda"], "data_venda")
        ensure_not_future(sale_date, "data_venda")
        lead_date = parse_date(lead["data_cadastro"], "data_cadastro")
        if sale_date < lead_date:
            raise AppError("Data da venda não pode ser menor que a data de cadastro do lead.")

        origem_fechamento = validate_enum(normalize_text(payload["origem_fechamento"]), SALES_ORIGINS, "origem_fechamento")

        meeting_id = payload.get("id_reuniao")
        if meeting_id:
            meeting = self.db.fetch_optional("select * from reuniao where id_reuniao = %s limit 1", [meeting_id])
            if not meeting:
                raise AppError("Reunião informada não existe.")
            if meeting["id_lead"] != lead["id_lead"]:
                raise AppError("Reunião informada deve pertencer ao mesmo lead da venda.")

        response = self.db.fetch_one(
            """
            insert into vendas (
              id_lead, id_servico, id_reuniao, origem_fechamento, valor_venda, data_venda
            )
            values (%s, %s, %s, %s, %s, %s)
            returning *
            """,
            [
                lead["id_lead"],
                payload["id_servico"],
                meeting_id,
                origem_fechamento,
                ensure_positive_number(payload["valor_venda"], "valor_venda"),
                sale_date.isoformat(),
            ],
        )
        self.lead_service.update_situation(lead["id_lead"], "Cliente")
        return response

    def update_sale(self, sale_id: int, payload: dict) -> dict:
        sale = self.get_one("id_venda", sale_id)
        immutable_messages = {
            "id_lead": "O lead da venda não pode ser alterado.",
            "id_servico": "O serviço da venda não pode ser alterado.",
            "origem_fechamento": "A origem da venda não pode ser alterada.",
            "valor_venda": "O valor da venda não pode ser alterado.",
            "data_venda": "A data da venda não pode ser alterada.",
            "id_reuniao": "A reunião vinculada da venda não pode ser alterada.",
        }

        for field, message in immutable_messages.items():
            if field in payload and payload[field] != sale.get(field):
                raise AppError(message)

        raise AppError("Nenhuma alteração permitida para vendas registradas.")

    def delete_sale(self, sale_id: int) -> None:
        sale = self.get_one("id_venda", sale_id)
        lead = self.lead_service.get_lead(sale["id_lead"])
        if canonical_text_key(lead["situacao"]) != canonical_text_key("Inativo"):
            raise AppError("Venda registrada só pode ser excluída quando o lead estiver Inativo.")
        self.db.execute("delete from vendas where id_venda = %s", [sale_id])
        self.lead_service.refresh_situation_from_history(sale["id_lead"])

File: backend/src/modules/sales/service.ts
import { SALES_ORIGINS } from "../../core/constants";
import { AppError } from "../../core/errors";
import { canonicalTextKey, normalizeText } from "../../core/formatters";
import { ensureNotFuture, ensurePositiveNumber, parseDate, requireFields, validateEnum } from "../../core/validators";
import { BaseService } from "../../services/base-service";
import { AttemptService } from "../attempts/service";
import { LeadService } from "../leads/service";

export class SaleService extends BaseService {
  protected tableName = "vendas";
  private leadService = new LeadService();
  private attemptService = new AttemptService();

  async listSales(filters: Record<string, string>) {
    let sql = "select * from vendas where 1=1";
    const params: unknown[] = [];

    if (filters.origem_fechamento) {
      sql += " and origem_fechamento = %s";
      params.push(this.resolveOrigin(filters.origem_fechamento));
    }
    if (filters.id_servico) {
      sql += " and id_servico = %s";
      params.push(Number(filters.id_servico));
    }
    if (filters.periodo_inicio) {
      sql += " and data_venda >= %s";
      params.push(filters.periodo_inicio);
    }
    if (filters.periodo_fim) {
      sql += " and data_venda <= %s";
      params.push(filters.periodo_fim);
    }

    sql += " order by data_venda desc, id_venda desc";
    return this.db.fetchAll(sql, params);
  }

  async createSale(payload: Record<string, unknown>, _userId: string) {
    return this.db.transaction(async () => {
      requireFields(payload, ["id_lead", "id_servico", "origem_fechamento", "valor_venda", "data_venda"]);
      const lead = (await this.leadService.getLead(Number(payload.id_lead))) as Record<string, unknown>;
      this.leadService.ensureAllowsNewRelatedRecords(lead);
      if (!(await this.attemptService.hasAttemptForLead(Number(lead.id_lead)))) {
        throw new AppError("Só é possível registrar venda após existir tentativa de contato.");
      }

      const service = await this.db.fetchOptional<Record<string, unknown>>(
        "select * from servicos where id_servico = %s limit 1",
        [Number(payload.id_servico)],
      );
      if (!service) {
        throw new AppError("Serviço informado não existe.");
      }

      const saleDate = parseDate(payload.data_venda, "data_venda");
      ensureNotFuture(saleDate, "data_venda");
      const leadDate = parseDate(lead.data_cadastro, "data_cadastro");
      if (saleDate < leadDate) {
        throw new AppError("Data da venda não pode ser menor que a data de cadastro do lead.");
      }

      const origemFechamento = this.resolveOrigin(String(payload.origem_fechamento));

      const meetingId = payload.id_reuniao ? Number(payload.id_reuniao) : null;
      if (meetingId) {
        const meeting = await this.db.fetchOptional<Record<string, unknown>>(
          "select * from reuniao where id_reuniao = %s limit 1",
          [meetingId],
        );
        if (!meeting) {
          throw new AppError("Reunião informada não existe.");
        }
        if (Number(meeting.id_lead) !== Number(lead.id_lead)) {
          throw new AppError("Reunião informada deve pertencer ao mesmo lead da venda.");
        }
      }

      const response = await this.db.fetchOne<Record<string, unknown>>(
        `
        insert into vendas (
          id_lead, id_servico, id_reuniao, origem_fechamento, valor_venda, data_venda
        )
        values (%s, %s, %s, %s, %s, %s)
        returning *
        `,
        [
          Number(lead.id_lead),
          Number(payload.id_servico),
          meetingId,
          origemFechamento,
          ensurePositiveNumber(payload.valor_venda, "valor_venda"),
          saleDate,
        ],
      );
      await this.leadService.updateSituation(Number(lead.id_lead), "Cliente");
      return response;
    });
  }

  async updateSale(saleId: number, payload: Record<string, unknown>) {
    return this.db.transaction(async () => {
      const sale = (await this.getOne("id_venda", saleId)) as Record<string, unknown>;
      const immutableMessages: Record<string, string> = {
        id_lead: "O lead da venda não pode ser alterado.",
        id_servico: "O serviço da venda não pode ser alterado.",
        origem_fechamento: "A origem da venda não pode ser alterada.",
        valor_venda: "O valor da venda não pode ser alterado.",
        data_venda: "A data da venda não pode ser alterada.",
        id_reuniao: "A reunião vinculada da venda não pode ser alterada.",
      };

      for (const [field, message] of Object.entries(immutableMessages)) {
        if (field in payload && payload[field] !== sale[field]) {
          throw new AppError(message);
        }
      }

      throw new AppError("Nenhuma alteração permitida para vendas registradas.");
    });
  }

  async deleteSale(saleId: number) {
    return this.db.transaction(async () => {
      const sale = (await this.getOne("id_venda", saleId)) as Record<string, unknown>;
      const lead = (await this.leadService.getLead(Number(sale.id_lead))) as Record<string, unknown>;
      if (canonicalTextKey(lead.situacao) !== canonicalTextKey("Inativo")) {
        throw new AppError("Venda registrada só pode ser excluída quando o lead estiver Inativo.");
      }
      await this.db.execute("delete from vendas where id_venda = %s", [saleId]);
      await this.leadService.refreshSituationFromHistory(Number(sale.id_lead));
    });
  }

  private resolveOrigin(value: string) {
    return validateEnum(normalizeText(value), SALES_ORIGINS, "origem_fechamento");
  }
}
